File: sdk/python/shc.py
"""
SHC (Self-Healing Distributed Cache) — Python Client SDK
Zero third-party dependencies (uses standard requests).
"""
import requests

class SHC:
    def __init__(self, host="http://13.127.44.111:8000"):
        self.host = host.rstrip("/")

    def get(self, key: str):
        """
        1.2ms in-memory RAM cache read.
        Automatically falls back to Amazon RDS PostgreSQL and auto-hydrates on cache misses.
        """
        res = requests.get(f"{self.host}/api/get", params={"key": key}, timeout=1.5)
        return res.json() if res.status_code == 200 else None

    def set(self, key: str, value: str, ttl_seconds: int = 300):
        """
        Write-Through write: synchronously saves in the 9-node distributed cache mesh
        and persists into Amazon RDS PostgreSQL with mathematical cascade recalculations.
        """
        payload = {"key": key, "value": value, "ttl_seconds": ttl_seconds}
        res = requests.post(f"{self.host}/api/set", json=payload, timeout=1.5)
        return res.json()

    def evict(self, key: str):
        """
        Invalidates key from RAM cache (leaves backing database records intact).
        """
        res = requests.delete(f"{self.host}/api/delete", params={"key": key}, timeout=1.5)
        return res.json()

    def trip(self, trip_id: str):
        """
        Queries NYC Taxi record by ID (e.g. 'trip:45210') with cache-aside acceleration.
        """
        res = requests.get(f"{self.host}/api/trip", params={"id": trip_id}, timeout=1.5)
        return res.json() if res.status_code == 200 else None

File: sdk/typescript/shc.ts
/**
 * SHC (Self-Healing Distributed Cache) — TypeScript / JavaScript Client SDK
 * Zero third-party dependencies (uses standard fetch API).
 */

export interface SHCResponse {
  status: "stored" | "hit" | "miss" | "deleted" | "error";
  key: string;
  value?: string;
  source?: "distributed_cache" | "backing_database";
  served_by?: string;
  is_failover?: boolean;
  db_updated?: boolean;
  message?: string;
}

export class SHC {
  private host: string;

  constructor(host = "http://13.127.44.111:8000") {
    this.host = host.replace(/\/$/, "");
  }

  /**
   * 1.2ms RAM read (auto-hydrates from RDS PostgreSQL on cache miss).
   */
  async get(key: string): Promise<SHCResponse | null> {
    const res = await fetch(`${this.host}/api/get?key=${encodeURIComponent(key)}`);
    return res.ok ? await res.json() : null;
  }

  /**
   * Write-Through: saves in distributed cache mesh and persists to Amazon RDS PostgreSQL.
   */
  async set(key: string, value: string, ttlSeconds = 300): Promise<SHCResponse> {
    const res = await fetch(`${this.host}/api/set`, {
      method: "POST",
      headers: { "Content-Type": "application/json" },
      body: JSON.stringify({ key, value, ttl_seconds: ttlSeconds }),
    });
    return await res.json();
  }

  /**
   * Evicts key from volatile in-memory RAM cache (leaves database row intact).
   */
  async evict(key: string): Promise<SHCResponse> {
    const res = await fetch(`${this.host}/api/delete?key=${encodeURIComponent(key)}`, {
      method: "DELETE",
    });
    return await res.json();
  }

  /**
   * Cache-Aside query against the 7.66M NYC Yellow Taxi dataset.
   */
  async trip(tripId: string): Promise<any> {
    const res = await fetch(`${this.host}/api/trip?id=${encodeURIComponent(tripId)}`);
    return res.ok ? await res.json() : null;
  }
}

export default SHC;
